packages/core-data/src/footnotes/get-footnotes-order.js

```diff
@@ -0,0 +1,30 @@
+/**
+ * Internal dependencies
+ */
+import getRichTextValuesCached from './get-rich-text-values-cached';
+
+const cache = new WeakMap();
+
+function getBlockFootnotesOrder( block ) {
+	if ( ! cache.has( block ) ) {
+		const content = getRichTextValuesCached( block ).join( '' );
+		const newOrder = [];
+
+		// https://github.com/WordPress/gutenberg/pull/43204 lands. We can then
+		// get the order directly from the rich text values.
+		if ( content.indexOf( 'data-fn' ) !== -1 ) {
+			const regex = /data-fn="([^"]+)"/g;
+			let match;
+			while ( ( match = regex.exec( content ) ) !== null ) {
+				newOrder.push( match[ 1 ] );
+			}
+		}
```

Comment:
given that we have easy access to a DOM here, why parse the HTML with the regular expressions? would it be just the same if we did this methodically instead? particularly since we're caching the values?

```js
if ( ! content.includes( 'data-fn' ) ) {
	cache.set( block, [] );
}

const dom = doTheStuffWeDoToParseHTML( content );
dom.querySelectorAll( '[data-fn]' ).foEach( ( node ) => {
	newOrder.push( node.getAttribute( 'data-fn' ) );
} )
```